user_sync/config.py

```diff
@@ -694,6 +611,138 @@ def describe_unused_values(self):
+    def load_from_yaml(filename, path_keys):
+        '''
+        loads a yaml file, processes the resulting dict to adapt values for keys
+        (the path to which is defined in PATH_KEYS) to a value that represents
+        a file reference relative to provided source file, and returns the
+        processed dict.
+        :type filename: str
+        :rtype dict
+        '''
+        try:
+            input_file = open(filename, 'r', 1)
+            yml = yaml.load(input_file)
+        except IOError as e:
+            # if a file operation error occurred while loading the
+            # configuration file, swallow up the exception and re-raise this
+            # as an configuration loader exception.
+            raise user_sync.error.AssertionException('Error reading configuration file: %s' % e)
+        except yaml.error.MarkedYAMLError as e:
+            # same as above, but indicate this problem has to do with
+            # parsing the configuration file.
+            raise user_sync.error.AssertionException('Error parsing configuration file: %s' % e)
+
+        dirpath = os.path.dirname(filename)
+    
+        def process_path_key(dictionary, keys, level, default_val):
+            '''
+            this function is used to process a single path key by replacing the
+            value for the key that is found into a path relative to the given
+            source file, with the assumption that the value is a file reference
+            to begin with. It is used recursively to navigate into child
+            dictionaries to search the path key.
+            type dictionary: dict
+            '''
+            def relative_path(filename):
+                '''
+                returns an absolute path that is resolved relative to the source
+                filename. The source filename is provided in the parent
+                load_from_yaml function, and os.path.abspath is used to return
+                the absolute path of the resolved relative path
+                type filename: str
+                rtype: str
+                '''
+                if dirpath and not os.path.isabs(filename):
+                    return os.path.abspath(os.path.join(dirpath, filename))
+                return filename
+            
+            def process_path_key_value(key):
+                '''
+                does the relative path processing for a single key of the
+                current dictionary
+                type key: str
+                '''
+                if dictionary.has_key(key):
+                    val = dictionary[key]
+                    if isinstance(val,str):
+                        dictionary[key] = relative_path(val)
+                    elif isinstance(val,list):
+                        paths = []
+                        for path in val:
+                            if isinstance(path, str):
+                                paths.append(relative_path(path))
+                        dictionary[key] = paths
```

Comment:
Almost there!  But this is excluding all the dict values in `val`.  Since `val` is going to be processed by `get_dict_from_sources`, it can be a list of pathnames and dictionaries, and you have to preserve the dictionaries.  So it should be
```python
elif isinstance(val, list):
    vals = []
    for entry in val:
        if isinstance(entry, str):
            vals.append(relative_path(path))
        else:
            vals.append(entry)
    dictionary[key] = vals